### PR TITLE
refactor!: Rename structs in Rust to rename classes in R

### LINKED ISSUE #149: Open discussion on renaming polars classes
I suggest to prefix all polars classes `Polars`, examples would be `PolarsDataFrame` `PolarsSeries` `PolarsDataType`

Pros:
 - Reduces risk of namespace collision with other R packages
 - Improves readability of rust code as currently native rust-polars struct and wrapper structs are named the same:
 
 readability example of current naming: rust-polars `LazyFrame` in a wrapper struct named `LazyFrame`. This can be a confusing to read, also the compiler error can be confusing.
image

Cons:
 - class/struct names become a bit long
 
 What naming do you (anyone) prefer/suggest? Should we rename? Prefixed RPolars?
 
 
Bonus info
 - Camel case is the only (unless explicitly disabled) allowed naming style for structs on rust side.
 - Currently, extendr can only name R side auto generated class the same as rust side wrapper structs.
 - Since r-polars only uses 2 structs + 1 class from rust-polars to public R, whereas py-polars use 2 structs + 2 classes, the final  public R side class name most be the same as the rust side wrapper struct. (**footnote)

(**the long footnote)
py-polars has 4 classes/structs from rust to python. DataFrame(rust-polars struct) -> PyDataFrame(wrapper struct) -> PyDataFrame (py03 internal class) -> DataFrame(py-polars public class). The vast majority of py-polars wrapper structs and wrapper classes are the thinnest possible.
r-polars only has 3 classes/struct from rust to python:  DataFrame(rust-polars struct) -> DataFrame(wrapper struct) -> DataFrame(extendr class). The R side DataFrame class though have both a set of public methods and a private method set. An early version of r-polars had 4 classes like py-polars and was using R6. That led to more conversions needed at runtime, more boiler plate code, and class instances was no longer just an R type ExternalPtr. Also, I felt it was sometimes cumbersome to make R6 behave exactly like py-polars syntax. Instead I decided to just implement extra base R S3 generic methods for the already auto-generated extendr-classes and add public method functions.

### PR DESCRIPTION
**Pull request by etiennebacher**
Close #149 

I just renamed the LazyFrame class (on R side) and struct (on rust side) to RPolarsLazyFrame. This compiles and works fine. We don't need to change the docs since this is only an internal (breaking) change.

@sorhawell @eitsupi if you agree with this then I can move forward and rename DataFrame and Expr

### CODE REVIEW

**Comment by sorhawell** [comment]
This looks very good! I would see it as minor change as some user may have written using class() or inherits() and there have to adjust the class name.

**Comment by sorhawell** [comment]
This looks very good! I would see it as minor change as some user may have written using class() or inherits() and there have to adjust the class name.

Funny thing @paleolimbot suggested this change very early in the project history, and I was like no that would be terrible. I finally came around to like it :)

**Comment by sorhawell** [comment]
This looks very good! I would see it as minor change as some user may have written using class() or inherits() and there have to adjust the class name.

Funny thing @paleolimbot suggested this change very early in the project history, and I was like no that would be terrible. I finally came around to like it :)


If you like you can also split it into a couple of PRs.

**Review by sorhawell — DISMISSED** [review]

**Comment by etiennebacher** [comment]
> I would see it as minor change as some user may have written using class() or inherits() and there have to adjust the class name.

Yes I'll put this in the NEWS

> If you like you can also split it into a couple of PRs.

I think I'll do all at once here

**Comment by eitsupi** [comment]
Thanks for working at this!
I'm definitely in favor of making all class name changes.

**Comment by etiennebacher** [comment]
I'm wondering if I should add the `RPolars` prefix to the other classes too, like `LazyGroupBy` or `DataType`:

https://github.com/pola-rs/r-polars/pull/554/files#diff-afa4237c90018088718e50ed6bf29c5d33bcf94ba925a9a5089683ba258e1343R270-R272

What do you think?

**Comment by eitsupi** [comment]
> I'm wondering if I should add the RPolars prefix to the other classes too

I think so.

File: src/rust/Cargo.toml [hunk 1]
```diff
@@ -127,6 +127,7 @@ features = [
   "search_sorted",
   "serde-lazy",
   "sign",
+  "sql",
```
Comment (eitsupi) [inline src/rust/Cargo.toml:130]: Is this intentional change?
The polars/sql feature is already included in the sql feature.

Comment (etiennebacher) [reply src/rust/Cargo.toml:130]: I thought it would fix the error in the docs CI, I can't reproduce it locally. I'll remove this.

Comment (eitsupi) [reply src/rust/Cargo.toml]: The error message seems come from here.
https://github.com/pola-rs/r-polars/blob/7bbe2c69fcad744a5442b4cff4ae4e071f2230d6/R/dataframe__frame.R#L74

IIUC, this line can changed to be `NULL`

Comment (etiennebacher) [reply src/rust/Cargo.toml]: Thanks, nice catch! Unfortunately there's still one error I can't track down though

**Comment by eitsupi** [comment]
I think we need update these.
https://github.com/pola-rs/r-polars/blob/7bbe2c69fcad744a5442b4cff4ae4e071f2230d6/R/zzz.R#L14-L27

See this one.
https://github.com/pola-rs/r-polars/blob/7bbe2c69fcad744a5442b4cff4ae4e071f2230d6/R/zzz.R#L104

**Comment by etiennebacher** [comment]
> See this one.

That works because you already used `RPolarsSQLContext$` in `extendr-wrappers.R`, cf my previous comment (that we posted at the same time)

**Comment by eitsupi** [comment]
> I haven't renamed all the `DataFrame$` in `extendr-wrappers.R`: https://github.com/pola-rs/r-polars/pull/554/files#diff-d9fa8df226b01f446a1391e800e573e396bdf7b99a514a2ffb2f4d1e9004c740
> 
> That would imply that we need to rename all `DataFrame_` functions to `RPolarsDataFrame_` (and similarly for all `LazyFrame_` and friends). I thought that would be unnecessary because we just want to change the class of the output. Should I replace all `DataFrame$` in `extendr-wrappers.R` too?

I'm sorry, but I don't know what you mean. The file is automatically generated, so you don't have to edit it manually, right?

My understanding is that they come from Rust struct names, so it's funny that currently it's not updated.

**Comment by eitsupi** [comment]
> That works because you already used `RPolarsSQLContext$` in `extendr-wrappers.R`, cf my previous comment (that we posted at the same time)

I haven't edited that file manually. It's just that the Rust struct name is used.

**Comment by etiennebacher** [comment]
I forgot this file is automatically generated, I always added the new methods by hand. My workflow is not the right one, but if I do `rextendr::document()` then `RPolarsSQLContext` doesn't appear in `extendr-wrappers.R`, even if I do `make build`, which is supposed to compile with full features, right?

**Comment by sorhawell** [comment]
> I think we may have to merge #516 first.

I have updated #516 , ready for review again

**Comment by etiennebacher** [comment]
Thanks for completing this PR @eitsupi

**Pull request merged**